<a href="https://colab.research.google.com/github/AlexK09/lab04/blob/master/Another_copy_of_Basic_of_C%2B%2B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!apt-get --purge remove cuda nvidia* libnvidia-*
!dpkg -l | grep cuda- | awk '{print $2}' | xargs -n1 dpkg --purge
!apt-get remove cuda-*
!apt autoremove
!apt-get update

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'nvidia-docker2' for glob 'nvidia*'
Note, selecting 'nvidia-cuda-toolkit-doc' for glob 'nvidia*'
Note, selecting 'nvidia-dkms-450-server' for glob 'nvidia*'
Note, selecting 'nvidia-headless-390' for glob 'nvidia*'
Note, selecting 'nvidia-cuda-toolkit-gcc' for glob 'nvidia*'
Note, selecting 'nvidia-headless-418' for glob 'nvidia*'
Note, selecting 'nvidia-headless-430' for glob 'nvidia*'
Note, selecting 'nvidia-headless-435' for glob 'nvidia*'
Note, selecting 'nvidia-headless-440' for glob 'nvidia*'
Note, selecting 'nvidia-headless-450' for glob 'nvidia*'
Note, selecting 'nvidia-headless-455' for glob 'nvidia*'
Note, selecting 'nvidia-headless-460' for glob 'nvidia*'
Note, selecting 'nvidia-headless-465' for glob 'nvidia*'
Note, selecting 'nvidia-headless-470' for glob 'nvidia*'
Note, selecting 'nvidia-dkms-470-server' for glob 'nvidia*'
Note, selecting 'nvidia-headless-495' f

In [ ]:
!apt install nvidia-cuda-toolkit
!apt install nvidia-cuda-dev

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  fonts-dejavu-core fonts-dejavu-extra libaccinj64-11.5 libatk-wrapper-java libatk-wrapper-java-jni
  libbabeltrace1 libcub-dev libcublas11 libcublaslt11 libcudart11.0 libcufft10 libcufftw10
  libcuinj64-11.5 libcupti-dev libcupti-doc libcupti11.5 libcurand10 libcusolver11 libcusolvermg11
  libcusparse11 libdebuginfod-common libdebuginfod1 libegl-dev libfontenc1 libgail-common libgail18
  libgl-dev libgl1-mesa-dev libgles-dev libgles1 libglvnd-core-dev libglvnd-dev libglx-dev
  libgtk2.0-0 libgtk2.0-bin libgtk2.0-common libipt2 libnppc11 libnppial11 libnppicc11 libnppidei11
  libnppif11 libnppig11 libnppim11 libnppist11 libnppisu11 libnppitc11 libnpps11 libnvblas11
  libnvidia-compute-495 libnvidia-compute-510 libnvidia-compute-525 libnvidia-ml-dev libnvjpeg11
  libnvrtc-builtins11.5 libnvrtc11.2 libnvtoolsext1 libnvvm4 libopengl-dev libq

In [ ]:
!nvcc --version


nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0


In [ ]:
!pip3 install git+https://github.com/andreinechaev/nvcc4jupyter.git


  Cloning https://github.com/andreinechaev/nvcc4jupyter.git to /tmp/pip-req-build-wz_2t3dn
  Running command git clone --filter=blob:none --quiet https://github.com/andreinechaev/nvcc4jupyter.git /tmp/pip-req-build-wz_2t3dn
  Resolved https://github.com/andreinechaev/nvcc4jupyter.git to commit 0a71d56e5dce3ff1f0dd2c47c29367629262f527
  Preparing metadata (setup.py) ... done
  Created wheel for NVCCPlugin: filename=NVCCPlugin-0.0.2-py3-none-any.whl size=4295 sha256=25a3d4e3fde5c4c94c4b711270ccd1483cf908c8f635569c0046185b26005497
  Stored in directory: /tmp/pip-ephem-wheel-cache-6e2ht8yv/wheels/a8/b9/18/23f8ef71ceb0f63297dd1903aedd067e6243a68ea756d6feea
Successfully built NVCCPlugin


In [ ]:
%load_ext nvcc_plugin


created output directory at /content/src
Out bin /content/result.out


In [ ]:
!ls /usr/local/cuda/

include  lib64	src  targets


In [ ]:
!ls /usr/bin | grep nv

dbus-update-activation-environment
deb-systemd-invoke
env
gencnval
iconv
makeconv
nvcc
nvdisasm
nvidia-bug-report.sh
nvidia-cuda-mps-control
nvidia-cuda-mps-server
nvidia-debugdump
nvidia-installer
nvidia-modprobe
nvidia-ngx-updater
nvidia-persistenced
nvidia-powerd
nvidia-settings
nvidia-sleep.sh
nvidia-smi
nvidia-uninstall
nvidia-xconfig
nvlink
nv-nsight-cu
nv-nsight-cu-cli
nvprof
nvprune
nvvp
piconv
preconv
printenv
uconv


In [ ]:
!git clone https://github.com/NVIDIA/cuda-samples

Cloning into 'cuda-samples'...
remote: Enumerating objects: 16830, done.
remote: Counting objects: 100% (2488/2488), done.
remote: Compressing objects: 100% (322/322), done.
remote: Total 16830 (delta 2211), reused 2190 (delta 2166), pack-reused 14342
Receiving objects: 100% (16830/16830), 133.20 MiB | 22.55 MiB/s, done.
Resolving deltas: 100% (14273/14273), done.
Updating files: 100% (3997/3997), done.


In [ ]:
%%cuda --name /content/cuda-samples/Samples/0_Introduction/simplePrintf/simplePrintf.cu --compile false

/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/11/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x1b): undefined reference to `main'
collect2: error: ld returned 1 exit status



In [2]:
!git clone https://github.com/welcheb/CUDA_examples.git

Cloning into 'CUDA_examples'...
remote: Enumerating objects: 40, done.
remote: Total 40 (delta 0), reused 0 (delta 0), pack-reused 40
Receiving objects: 100% (40/40), 10.53 MiB | 15.98 MiB/s, done.
Resolving deltas: 100% (10/10), done.


In [3]:
!nvcc /content/CUDA_examples/devicequery2.cu

/content/CUDA_examples/devicequery2.cu: In function ‘int main(int, char**)’:
/content/CUDA_examples/devicequery2.cu:20:15: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘size_t’ {aka ‘long unsigned int’} []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=]8;;]
   20 |         printf("totalGlobalMem: %d B = %d MB\n", deviceProp.totalGlobalMem, deviceProp.totalGlobalMem / 1048576);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                            |
      |                                                            size_t {aka long unsigned int}
/content/CUDA_examples/devicequery2.cu:20:15: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘size_t’ {aka ‘long unsigned int’} []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=]8;;]
   20 |         printf("totalGlobalMem: %d B = %

In [4]:
!./a.out

deviceCount: 1
name: Tesla T4
major: 7
minor: 5
multiProcessorCount: 40
totalGlobalMem: -1344471040 B = 15101 MB
sharedMemPerBlock: 49152 B = 48 KB
totalConstMem: 65536 B = 64 KB
regsPerBlock: 65536
ECCEnabled: 1
kernelExecTimeoutEnabled: 0
clockRate: 1590000 KHz = 1590 MHz
memoryClockRate: 5001000 KHz = 5001 MHz
memoryBusWidth: 256 bits
l2CacheSize: 4194304 B = 4096 KB
warpSize: 32
maxThreadsPerMultiProcessor: 1024
maxThreadsPerBlock: 1024
maxThreadsDim[0]: 1024
maxThreadsDim[1]: 1024
maxThreadsDim[2]: 64
maxGridSize[0]: 2147483647
maxGridSize[1]: 65535
maxGridSize[2]: 65535
deviceOverlap: 1
asyncEngineCount: 3
integrated: 0
canMapHostMemory: 1
concurrentKernels: 1
tccDriver: 0
unifiedAddressing: 1
pciBusID: 0
pciDeviceID: 4
computeMode: 0
computeMode: Default (multiple host threads can use ::cudaSetDevice() with device simultaneously)


In [5]:
!nvcc /content/CUDA_examples/devicequery.cu -o script

/content/CUDA_examples/devicequery.cu: In function ‘void printDevProp(cudaDeviceProp)’:
/content/CUDA_examples/devicequery.cu:9:8: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘size_t’ {aka ‘long unsigned int’} []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=]8;;]
    9 |     printf("Total global memory:           %u\n",  devProp.totalGlobalMem);
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~
      |                                                       |
      |                                                       size_t {aka long unsigned int}
/content/CUDA_examples/devicequery.cu:10:8: warning: format ‘%u’ expects argument of type ‘unsigned int’, but argument 2 has type ‘size_t’ {aka ‘long unsigned int’} []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wformat=-Wformat=]8;;]
   10 |     printf("Total shared memory per block: %u\n",  devProp.sharedMemP

In [7]:
!./script

CUDA Device Query...
There are 1 CUDA devices.

CUDA Device #0
Major revision number:         7
Minor revision number:         5
Name:                          Tesla T4
Total global memory:           2950496256
Total shared memory per block: 49152
Total registers per block:     65536
Warp size:                     32
Maximum memory pitch:          2147483647
Maximum threads per block:     1024
Maximum dimension 0 of block:  1024
Maximum dimension 1 of block:  1024
Maximum dimension 2 of block:  64
Maximum dimension 0 of grid:   2147483647
Maximum dimension 1 of grid:   65535
Maximum dimension 2 of grid:   65535
Clock rate:                    1590000
Total constant memory:         65536
Texture alignment:             512
Concurrent copy and execution: Yes
Number of multiprocessors:     40
Kernel execution timeout:      No


In [8]:
!nvcc /content/CUDA_examples/sumArraysOnGPU-small-case.cu -o big_script

In [9]:
!./big_script

./big_script Starting...
Vector size 32
Execution configuration <<<32, 1>>>
Arrays match.



In [10]:
!nvcc /content/CUDA_examples/sumArraysOnGPU-timer.cu -o big_timer

In [11]:
!./big_timer

./big_timer Starting...
Using Device 0: Tesla T4
Vector size 16777216
sumArraysOnHost Time elapsed 0.069693 sec
sumArraysOnGPU <<<16384, 1024>>> Time elapsed 0.000877 sec
GPU speedup factor = 79.45 sec
Arrays match.



In [12]:
!nvcc /content/CUDA_examples/hello.cu -o hello

In [13]:
!./hello

Hello World from CPU!
Hello World from GPU! threadIdx.x=0
Hello World from GPU! threadIdx.x=1
Hello World from GPU! threadIdx.x=2
Hello World from GPU! threadIdx.x=3
Hello World from GPU! threadIdx.x=4
Hello World from GPU! threadIdx.x=5
Hello World from GPU! threadIdx.x=6
Hello World from GPU! threadIdx.x=7
Hello World from GPU! threadIdx.x=8
Hello World from GPU! threadIdx.x=9


In [14]:
!git clone https://bitbucket.org/VladyslavGorbatiuk/cuda-parallel-programming-code.git

Cloning into 'cuda-parallel-programming-code'...
remote: Enumerating objects: 55, done.
remote: Counting objects: 100% (55/55), done.
remote: Compressing objects: 100% (54/54), done.
remote: Total 55 (delta 18), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (55/55), 14.53 KiB | 82.00 KiB/s, done.


In [31]:
!mkdir tmp && cd /content/tmp && cmake /content/cuda-parallel-programming-code/

CMake Deprecation Warning at CMakeLists.txt:1 (CMAKE_MINIMUM_REQUIRED):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:10 (INCLUDE):
  Policy CMP0146 is not set: The FindCUDA module is removed.  Run "cmake
  --help-policy CMP0146" for policy details.  Use the cmake_pol

In [32]:
!cd tmp && make

[ 25%] Building NVCC (Device) object Part2/CMakeFiles/Part2CUDA.dir/Part2CUDA_generated_CUDA.cu.o
[ 50%] Linking CXX static library libPart2CUDA.a
[ 50%] Built target Part2CUDA
[ 75%] Building CXX object Part2/CMakeFiles/blur_image.dir/main.cpp.o
[100%] Linking CXX executable ../bin/blur_image
[100%] Built target blur_image


In [36]:
!./tmp/bin/blur_image /content/how-to-convert-low-to-high-resolution-in-photoshop-32.jpg

OpenCV time (ms):5
OpenMP time (ms):1180
OpenMP similarity:0.00171731
CUDA time kernel (ms): 0.522688
CUDA time full (ms):136
CUDA similarity:0.00320056
CUDA time kernel (ms): 0.517184
CUDA time full (ms):2
CUDA similarity:0.00170965
CUDA time kernel (ms): 0.51712
CUDA time full (ms):2
CUDA similarity:0.00170965
CUDA time kernel (ms): 0.516064
CUDA time full (ms):2
CUDA similarity:0.00170965


#**Introduction**

C++ (pronounced see plus plus) is a general purpose programming language that is free-form and compiled. It is regarded as an intermediate-level language, as it comprises both high-level and low-level language features. It provides imperative, object-oriented and generic programming features.

C++ is one of the most popular programming languages and is implemented on a wide variety of hardware and operating system platforms. As an efficient performance driven programming language it is used in systems software, application software, device drivers, embedded software, high-performance server and client applications, and entertainment software such as video games. Various entities provide both open source and proprietary C++ compiler software, including the FSF, LLVM, Microsoft and Intel

**What is C++**
C++ is a cross-platform language that can be used to create high-performance applications.

**Who developed C++**
C++ was developed by Bjarne Stroustrup.



#1- **Hellow Words**
Every C++ program uses libraries, which give the ability to execute necessary functions. For example, the most basic function called 'cout', which prints to the screen, is defined in the iostream header file.

To add the ability to run the cout command to our program, we must add following include directive to our first line of the code:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
  cout << "Goodbye, World!" << std::endl;
  return 0;
}

FileNotFoundError: ignored

#2-**Comments**
Comments can be used to explain C++ code, and to make it more readable. It can also be used to prevent execution when testing alternative code. Comments can be singled-lined or multi-lined.

Single-line comments start with two forward slashes (//).

Any text between // and the end of the line is ignored by the compiler (will not be executed).

This example uses a single-line comment before a line of code:





In [ ]:
%%cu
/* hellow wors
sdfdf
*/
#include <iostream>
using namespace std; // it used for explaination

int main() {
  cout << "Goodbye, World!" << std::endl;
  return 0;
}

Goodbye, World!



#3-**C++ User Input**

.You have already learned that cout is used to output (print) values. Now we will use cin to get user input.

.cin is a predefined variable that reads data from the keyboard with the extraction operator (>>).

In [ ]:
%%cu

#include <iostream>
using namespace std; // it used for explaination

int main() {
  int x=0;
cout << "Type a number: "; // Type a number and press enter
cin >> x; // Get user input from the keyboard
cout << "Your number is: " << x; // Display the input value
}

Type a number: Your number is: 0


#4-**Variable and types**

**Data types**

C++ provides a rich set of built-in as well as user defined data types. Following table lists down seven basic C++ data types:

**Built-in**
- Boolean - boolean - either true or false (Size 4 bytes).
- Integers - whole numbers which can be both positive and negative. Defined using char, int.(Size 4 bytes)
- Floating point numbers - real numbers (numbers with - fractions). Defined using float and double (Size 4 bytes)
- Double - Stores Fractional numbers.Containing one or more decimals (size 8 bytes )
- Char - Stores a Single Characters
Valueless using the void keyword
Wide character using the wchar_t keyword(Size 1 bytes)

**Typdefs**
Typedefs allow for creating new names (think of them as aliases) for existing types. Following is the simple syntax to define a new type using typedef:

**Defining variables**
For numbers, we will usually use the type int, which an integer in the size of a "word" the default number size of the machine which your program is compiled on. On most computers today, it is a 32-bit number, which means the number can range from -2,147,483,648 to 2,147,483,647 (same as long).

To define the variables foo and bar, we need to use the following syntax:

int foo;

int bar = 1;

The variable foo can be used, but since we did not initialize it, we don't know what's in it. The variable bar contains the number 1.

Now, we can do some math. Assuming a, b, c, d, and e are variables, we can simply use plus, minus and multiplication operators in the following notation, and assign a new value to a:

int a = 0, b = 1, c = 2, d = 3, e = 4;

a = b - c + d * e;

cout << a << endl; // will print 1-2+3*4 = 11



In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
  int a = 3;
  float b = 4.5;
  double c = 5.25;
  double sum;

  /* Your code goes here */
    sum =a+b+c;

  cout << "The sum of a, b, and c is" << sum << endl;
  return 0;
}


The sum of a, b, and c is12.75



In [ ]:
%%cu
#include <iostream>
using namespace std;

int main () {
  int myNum = 1000;
  cout << myNum;
  return 0;
}



1000


In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    float mynumb=2.25;
    cout <<mynumb;
    return 0;
}


2.25


In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    double mynumb=66.87;
    cout <<mynumb;
    return 0;

}


In [ ]:
%%cu
#include <iostream>
using namesapace std;

int main() {
  int x = 5;
  int y = 3;
  cout << x + y;
  return 0;
}



##4.1**-Strings**
Strings are used for storing text.

A string variable contains a collection of characters surrounded by double quotes


In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    string shahzad = "hello";
    cout << shahzad;
    return 0;
}


hello


# 5-**C++ Operators**
Operators are used to perform operations on variables and values.

In the example below, we use the + operator to add together two values:

## 5.1**Arithmetic Operators**
Arithmetic operators are used to perform common mathematical operations.

.(+)  (Addition) 	(Adds together two values)  	(x + y)

.(-) 	(Subtraction)	(Subtracts one value from another)	(x - y)

.(*) (Multiplication) 	(Multiplies two values) 	(x * y)

.(/) 	(Division) 	(Divides one value by another) 	(x / y)

.(%) 	(Modulus) 	(Returns the division remainder) 	(x % y)

.(++) 	(Increment) 	(Increases the value of a variable by 1) 	(++x)

.(--) 	(Decrement) 	(Decreases the value of a variable by 1) 	(--x)

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int a = 7;
    int b = 6;
    int resultplus;
    resultplus=a+b;
    int resultminus;
    resultminus=a-b;
    int resultmultiplaction;
    resultmultiplaction=a*b;
    int resultdivision;
    resultdivision=a/b;
    cout <<"additiona+b="<<resultplus<<endl;
    cout <<"minusa-b="<<resultminus<<endl;
    cout <<"resultmultiplactiona*b="<<resultmultiplaction<<endl;
    cout <<"divisiona/b="<<resultdivision<<endl;
    return 0;
}


#**5.2Assignment Operators**

Assignment operators are used to assign values to variables.

In the example below, we use the assignment operator (=) to assign the value 10 to a variable called x:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int a; // this is definition of assignment
    a = 2; //
    cout <<a<<endl;
    return 0;
}

## 5.3**Comparison Operators**

Comparison operators are used to compare two values.

Note: The return value of a comparison is either true (1) or false (0).

In the following example, we use the greater than operator (>) to find out if 5 is greater than 3:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int a =5;
    int b =4;
    cout <<(a>b)<<endl;
    return 0;
}


1



##5.4**Logical Operators**
Logical operators are used to determine the logic between variables or values:



In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int a=3;
    int b=5;
    cout <<(a>3 && a<7)<<endl; // returns true (1) because 3 is greater than 2 AND 2 is less than 3
    cout <<(b>3 || b<4)<<endl; // returns true (1) because one of the conditions are true (5 is greater than 3, but 5 is not less than 4)
    return 0;
}


#6-**C++ Math**
C++ has many functions that allows you to perform mathematical tasks on numbers.


##6.1-**Max and min**
The max(x,y) function can be used to find the highest value of x and y:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int a;
    a = 2;
    int b;
    b = 5;
    cout <<max(a,b)<<endl; // this function is used to find max value between two
    // value
    cout <<min(a,b)<<endl; // this function is used to find min values beteen
   // two  value
    return 0;
}


##6.2-**math Header**
Other functions, such as sqrt (square root), round (rounds a number) and log (natural logarithm), can be found in the <cmath> header file:


In [ ]:
%%cu
#include <iostream>
#include <cmath>
using namespace std;

int main(){
    int a;
    a = 4;
    int b;
    b = -2;
    cout <<"square value of a="<<sqrt(a)<<endl; // sqrt function is used to find squar root of varia
    cout <<"absolate vale of b="<<abs(b)<<endl;
}




#7-**C++ Booleans**
Very often, in programming, you will need a data type that can only have one of two values, like:

YES / NO

ON / OFF

TRUE / FALSE


For this, C++ has a bool data type, which can take the values true (1) or false (0).



In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    bool a =true;
    bool b =false;
    cout <<a<<endl;
    cout <<b<<endl;
    return 0;
}


1
0



##**7.1-Boolean Expression**
A Boolean expression is a C++ expression that returns a boolean value: 1 (true) or 0 (false).

You can use a comparison operator, such as the greater than (>) operator to find out if an expression (or a variable) is true:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
           int a;
           a = 2;
           int b;
           b = 1;
           cout <<"true and false"<<(a>b)<<endl;
           return 0;
}


true and false1



#**8-C++ Conditions and If Statements**
C++ supports the usual logical conditions from mathematics:

- Less than: a < b
- Less than or equal to: a <= b
- Greater than: a > b
- Greater than or equal to: a >= b
- Equal to a == b
- Not Equal to: a != b
- You can use these conditions to perform different actions for different decisions.

C++ has the following conditional statements:

- Use if to specify a block of code to be executed, if a specified condition is true
- Use else to specify a block of code to be executed, if the same condition is false
- Use else if to specify a new condition to test, if the first condition is false
- Use switch to specify many alternative blocks of code to be executed


##**8.1-The if Statement**
Use the if statement to specify a block of C++ code to be executed if a condition is true.

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    if (5>10) {
        cout <<"20 is greater than 10"<<endl;
    }
    cout <<"main body exution"<<endl;
    return 0;
}


##**-8.2The else Statement**
Use the else statement to specify a block of code to be executed if the condition is false.

In [ ]:
#include <iostream>
using namespace std;

int main() {
  int time = 20;
  if (time < 18) {
    cout << "Good day.";
  } else {
    cout << "Good evening.";
  }
  return 0;
}


##**8.3-The else if Statement**
Use the else if statement to specify a new condition if the first condition is false.

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
  int time = 22;
  if (time < 10) {
    cout << "Good morning.";
  } else if (time < 20) {
    cout << "Good day.";
  } else {
    cout << "Good evening.";
  }
  return 0;
}


#**9-C++ Switch Statements**
Use the switch statement to select one of many code blocks to be executed.
This is how it works:
-  The switch expression is evaluated once
- The value of the expression is compared with the values of each case
- If there is a match, the associated block of code is executed
- The break and default keywords are optional, and will be described later in this chapter

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int day=9;
    switch(day){
     case 1:
     cout <<"monday";
     break;
     case 2:
     cout <<"tuesday";
     break;
     default:
     cout <<"no case matches";
    }
    return 0;
}


**The break Keyword**

When C++ reaches a break keyword, it breaks out of the switch block.

This will stop the execution of more code and case testing inside the block.

When a match is found, and the job is done, it's time for a break. There is no need for more testing.

A break can save a lot of execution time because it "ignores" the execution of all the rest of the code in the switch block.

**The default Keyword**

The default keyword specifies some code to run if there is no case match:

#**10-Loop**
Loops can execute a block of code as long as a specified condition is reached.

Loops are handy because they save time, reduce errors, and they make code more readable.

**While Loop**

The while loop loops through a block of code as long as a specified condition is true:


In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int i =0;
    while (i<5){
        cout <<"shahzad"<<"\n";
        i++;
    }
    return 0;
}


##**10.1-The Do/While Loop**
 - The do/while loop is a variant of the while loop.
 -This loop will execute the code block once, before checking if the condition is true, then it will repeat the loop as long as the condition is true.

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    int i =9;
    do{
        cout <<"shahzad"<<"\n";
       i++;
    }
    while(i<5);
    return 0;
}


##**10.2-C++ Break**
The break statement can also be used to jump out of a loop.

This example jumps out of the loop when i is equal to 4:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
     for (int i = 0; i < 14; i++) {
         if (i == 8){
          break;
         }
    cout << i << "\n";
  }
  return 0;
}



##**10.3-C++ Continue**

The continue statement breaks one iteration (in the loop), if a specified condition occurs, and continues with the next iteration in the loop.

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main(){
    for (int i = 0; i < 12; i++) {
        if (i == 7){
            continue;
        }
        cout <<i<<"\n";
    }
    return 0;
}


#**11-C++ Arrays**
Arrays are used to store multiple values in a single variable, instead of declaring separate variables for each value.

To declare an array, define the variable type, specify the name of the array followed by square brackets and specify the number of elements it should store:

**string cars[4];**

To create an array of three integers, you could write:
**int myNum[3] = {10, 20, 30};**

**Access the Elements of an Array**

You access an array element by referring to the index number.

This statement accesses the value of the first element in cars:

string cars[4] = {"Volvo", "BMW", "Ford", "Mazda"};
cout << cars[0];
// Outputs Volvo

In [ ]:
%%cu
#include <iostream>
#include <string>
using namespace std;

int main() {
    string Ali[4] = {"ijaz","hussain", "book", "work"};
    cout <<Ali[0];
    return 0;
}


**Change an Array Element**

To change the value of a specific element, refer to the index number:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
    string hussain[4] = {"ali","abbas","summar","ayan"};
    hussain[0]="haya";
    cout <<hussain[0];
    return 0;
}

##**11.1-C++ Arrays and Loops**

You can loop through the array elements with the for loop.

The following example outputs all elements in the cars array:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
    int i;
    string ali[4] = {"shahzad","sohail","hamza","haya"};
    for( i = 0; i < 4; i++) {
        cout << i << ":" <<ali[i] << "\n";
    }
    return 0;
}

UsageError: Cell magic `%%cu` not found.


## **11.2-C++ Omit Array Size**

You don't have to specify the size of the array. But if you don't, it will only be as big as the elements that are inserted into it:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
    int i;
    string ali[] = {"hello", "world","key","school"};
    for(i =0; i < 4; i++) {
        cout << i << ":" <<ali[i] << "\n";
    }
    return 0;
}


UsageError: Cell magic `%%cu` not found.


#**12-C++ References**

 A reference variable is a "reference" to an existing variable, and it is created with the & operator:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
    string ali = "pizza";
    string &meal = ali;

    cout << ali <<endl;
    cout << meal <<endl;
    return 0;
}


#**13-C++ Memory Address**

In the example from the previous page, the & operator was used to create a reference variable. But it can also be used to get the memory address of a variable; which is the location of where the variable is stored on the computer.

When a variable is created in C++, a memory address is assigned to the variable. And when we assign a value to the variable, it is stored in this memory address.

To access it, use the & operator, and the result will represent where the variable is stored:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int main() {
    string food = "pizza";

    cout <<&food;
    return 0;
}


#**14-C++ Pointers**

You learned from the previous chapter, that we can get the memory address of a variable by using the & operator:

A pointer however, is a variable that stores the memory address as its value.


A pointer variable points to a data type (like int or string) of the same type, and is created with the * operator. The address of the variable you're working with is assigned to the pointer:

**Example explained**

Create a pointer variable with the name ptr, that points to a string variable, by using the asterisk sign * (string* ptr). Note that the type of the pointer has to match the type of the variable you're working with.

Use the & operator to store the memory address of the variable called food, and assign it to the pointer.

Now, ptr holds the value of food's memory address.

In [ ]:
%%cu
#include <iostream>
using namespace std;
 int main(){
     string food = "pizza";
     string* ptr = &food;

     cout << food <<endl;
     cout << &food <<endl;
     cout << ptr <<endl;
     return 0;
     }


#**15-C++ Functions**

A function is a block of code which only runs when it is called.

You can pass data, known as parameters, into a function.

Functions are used to perform certain actions, and they are important for reusing code: Define the code once, and use it many times.

**Create a Function**

C++ provides some pre-defined functions, such as main(), which is used to execute code. But you can also create your own functions to perform certain actions.

To create (often referred to as declare) a function, specify the name of the function, followed by parentheses ():

**Syntax**

void myFunction() {
  // code to be executed
}

**Example Explained**

- myFunction() is the name of the function
- void means that the function does not have a return  value. You will learn more about return values later in the next chapter
- inside the function (the body), add code that defines what the function should do

**Call a Function**

Declared functions are not executed immediately. They are "saved for later use", and will be executed later, when they are called.

To call a function, write the function's name followed by two parentheses () and a semicolon ;

In the following example, myFunction() is used to print a text (the action), when it is called:

**Example**

Inside main, call myFunction():




In [ ]:
%%cu
#include <iostream>
using namespace std;

void pakistan() {
    cout << "long live pakistan!";
}
int main(){
    cout <<"first output"<<endl;
  pakistan();
  cout <<"second output"<<endl;
   pakistan();
   return 0;
}


**Function Declaration and Definition**

A C++ function consist of two parts:

- Declaration: the function's name, return type, and parameters (if any)
- Definition: the body of the function (code to be executed)

Note: If a user-defined function, such as myFunction() is declared after the main() function, an error will occur. It is because C++ works from top to bottom; which means that if the function is not declared above main(), the program is unaware of it:


In [ ]:
%%cu
#include <iostream>
using namespace std;

void myFunction() {
  cout << "I just got executed!";
}


int main() {
  myFunction();
  return 0;
}


##**15.1-C++ Function Parameters**

Information can be passed to functions as a parameter. Parameters act as variables inside the function.

Parameters are specified after the function name, inside the parentheses. You can add as many parameters as you want, just separate them with a comma:

When a parameter is passed to the function, it is called an argument. So, from the example above: fname is a parameter, while Liam, Jenny and Anja are arguments.

In [ ]:
%%cu
#include <iostream>
using namespace std;

void hussain(string word) {
    cout << word <<endl;
}
    int main(){
        hussain("school");
        hussain("book");
        hussain("pen");
        hussain("pencial");
        return 0;
    }

##**15.2-C++ Default Parameters**

You can also use a default parameter value, by using the equals sign (=).

If we call the function without an argument, it uses the default value ("Norway"):

A parameter with a default value, is often known as an "optional parameter". From the example above, country is an optional parameter and "Norway" is the default value.

In [ ]:
%%cu
#include <iostream>
using namespace std;

void shahzad(string name ="haider"){
cout << name <<endl;
}
int main(){
shahzad("haya");
shahzad("summar");
shahzad();
shahzad("ayan");
return 0;
}


##**15.3-C++ Multiple Parameters**

Inside the function, you can add as many parameters as you want:

Note that when you are working with multiple parameters, the function call must have the same number of arguments as there are parameters, and the arguments must be passed in the same order.

In [ ]:
%%cu
#include <iostream>
using namespace std;

void work(string shahzad, int age) {
     cout << shahzad << " endl. " << age << " years old.endl";
}
int main(){
    work("haya",5);
    work("summar",11);
    work("mahdi",4);
    work("lahar",10);
    return 0;
}


##**15.4-C++ The Return Keyword**

The void keyword, used in the previous examples, indicates that the function should not return a value. If you want the function to return a value, you can use a data type (such as int, string, etc.) instead of void, and use the return keyword inside the function:

In [ ]:
%%cu
#include <iostream>
using namespace std;

int world(int x, int y) {
    int sum = x+y;
    return sum;
}
int main(){
    int finalsum = world(5 , 7);
    cout <<finalsum<<endl;
    return 0;
}

12



##**15.5-C++ Function Overloading**

With function overloading, multiple functions can have the same name with different parameters:

Instead of defining two functions that should do the same thing, it is better to overload one.

In the example below, we overload the plusFunc function to work for both int and double:

Note: Multiple functions can have the same name as long as the number and/or type of parameters are different.

In [ ]:
%%cu
#include <iostream>
using namespace std;

int myfunction(int x, int y) {
    return  x + y;
}
double myfunction(double x, double y) {
    return x + y;
}
int main(){
    int number1 = myfunction(8 , 6);
    double number2 = myfunction(5.5 , 2.5);
    cout <<"Int: " << number1<<endl;
    cout <<"Double: " <<number2<<endl;
    return 0;
}

Int: 14
Double: 8



#**16-C++ OOP**

**C++ What is OOP?**

OOP stands for Object-Oriented Programming.

Procedural programming is about writing procedures or functions that perform operations on the data, while object-oriented programming is about creating objects that contain both data and functions.

Object-oriented programming has several advantages over procedural programming:

- OOP is faster and easier to execute
- OOP provides a clear structure for the programs
- OOP helps to keep the C++ code DRY "Don't Repeat Yourself", and makes the code easier to maintain, modify and debug
- OOP makes it possible to create full reusable applications with less code and shorter development time.


##**16.1-C++ What are Classes and Objects?**

Classes and objects are the two main aspects of object-oriented programming.

Look at the following illustration to see the difference between class and objects:

So, a class is a template for objects, and an object is an instance of a class.

When the individual objects are created, they inherit all the variables and functions from the class.

You will learn much more about classes and objects in the next chapter.

![](https://drive.google.com/uc?export=view&id=1OiI0OM8SZDffAmtpczf56BYziR6f_b2T)

##**16.2-C++ Classes and Objects**

 C++ is an object-oriented programming language.

Everything in C++ is associated with classes and objects, along with its attributes and methods. For example: in real life, a car is an object. The car has attributes, such as weight and color, and methods, such as drive and brake.

Attributes and methods are basically variables and functions that belongs to the class. These are often referred to as "class members".

A class is a user-defined data type that we can use in our program, and it works as an object constructor, or a "blueprint" for creating objects.

**Create a Class**
To create a class, use the class keyword:
Create a class called "MyClass":


In [ ]:
class MyClass {       // The class
  public:             // Access specifier
    int myNum;        // Attribute (int variable)
    string myString;  // Attribute (string variable)
};

**Example explained**
- The class keyword is used to create a class called MyClass.
- The public keyword is an access specifier, which specifies that members (attributes and methods) of the class are accessible from outside the class. You will learn more about access specifiers later.

- Inside the class, there is an integer variable myNum and a string variable myString. When variables are declared within a class, they are called attributes.

- At last, end the class definition with a semicolon ;.

##**16.3-Create an Object**

In C++, an object is created from a class. We have already created the class named MyClass, so now we can use this to create objects.

To create an object of MyClass, specify the class name, followed by the object name.

To access the class attributes (myNum and myString), use the dot syntax (.) on the object:

**Example**
Create an object called "myObj" and access the attributes:


In [ ]:
%%cu
#include <iostream>
using namespace std;

class MyClass {       //  class
  public:             //  specifier
    int myNum;        // Attribute (int )
    string myString;  // Attribute (string)
};

int main() {
  MyClass myObj;  //  object of MyClass

  // Access attributes and set values
  myObj.myNum = 70;
  myObj.myString = "pakistan";

  // Print values
  cout << myObj.myNum <<endl;
  cout << myObj.myString <<endl;
  return 0;
}


##**16.4-Multiple Objects**
You can create multiple objects of one class:

In [ ]:
%%cu
#include <iostream>
using namespace std;

class car {
    public:
    string brand;
    string model;
    int year;
};

int main(){
    car carobj1;
    carobj1.brand = "BMW";
    carobj1.model = "vvt";
    carobj1.year = 2002;

    car carobj2;
    carobj2.brand = "toyota";
    carobj2.model = "vvtr";
    carobj2.year = 2006;

    cout << carobj1.brand << " " << carobj1.model << " " << carobj1.year <<endl;
    cout << carobj2.brand << " " <<carobj2.model << " " << carobj2.year <<endl;
    return 0;
}

##**16.5-C++ Class Methods**
Methods are functions that belongs to the class.

There are two ways to define functions that belongs to a class:

- Inside class definition
- Outside class definition

In the following example, we define a function inside the class, and we name it "myMethod".

**Note**: You access methods just like you access attributes; by creating an object of the class and by using the dot syntax (.):

**inside**

In [ ]:
%%cu
#include <iostream>
using namespace std;

class pakistan {
    public:
    void mymethod() {
        cout <<"pakistani people";
    }
    };
    int main() {
        pakistan myobj;
        myobj.mymethod();
        return 0;
    }


pakistani people


**0utside example**

In [ ]:
%%cu
#include <iostream>
using namespace std;

class myclass {
    public:
    void mymethod();
};
void myclass::mymethod(){
    cout << "pakistan";
}
 int main(){
     myclass myobj;
     myobj.mymethod();
     return 0;
 }

#**17-C++ Constructors**

A constructor in C++ is a special method that is automatically called when an object of a class is created.

To create a constructor, use the same name as the class, followed by parentheses ():

In [ ]:
%%cu
#include <iostream>
using namespace std;

class shahzad {     // class
  public:           //  specifier
    shahzad() {     // Constructor
      cout << "Hello World!";
    }
};

int main() {
  shahzad myObj;    // this will call the constructor
  return 0;
}


UsageError: Cell magic `%%cu` not found.


#**17.1-Constructor Parameters**

Constructors can also take parameters (just like regular functions), which can be useful for setting initial values for attributes.

The following class have brand, model and year attributes, and a constructor with different parameters. Inside the constructor we set the attributes equal to the constructor parameters (brand=x, etc). When we call the constructor (by creating an object of the class), we pass parameters to the constructor, which will set the value of the corresponding attributes to the same:

In [ ]:
%%cu
#include <iostream>
using namespace std;

class Car {        //  class
  public:          //  specifier
    string brand;  // Attribute
    string model;  // Attribute
    int year;      // Attribute
    Car(string x, string y, int z) {  // Constructor with parameters
      brand = x;
      model = y;
      year = z;
    }
};

int main() {
  // Create Car objects and call the constructor with different values
  Car carObj1("toyota", "vvt", 2006);
  Car carObj2("honda", "vvtr", 2012);

  // Print values
  cout << carObj1.brand << " " << carObj1.model << " " << carObj1.year << endl;
  cout << carObj2.brand << " " << carObj2.model << " " << carObj2.year << endl;
  return 0;
}


#**18-C++ Access Specifiers**
By now, you are quite familiar with the public keyword that appears in all of our class examples:

The public keyword is an access specifier. Access specifiers define how the members (attributes and methods) of a class can be accessed. In the example above, the members are public - which means that they can be accessed and modified from outside the code.

However, what if we want members to be private and hidden from the outside world?

In C++, there are three access specifiers:

- public - members are accessible from outside the class
- private - members cannot be accessed (or viewed) from outside the class
- protected - members cannot be accessed from outside the class, however, they can be accessed in inherited classes. You will learn more about Inheritance later.

In the following example, we demonstrate the differences between public and private members:

In [ ]:
%%cu
#include <iostream>
using namespace std;

class myclass{     // class
    public:        // public acccess specifier
    int x;         // attributes
    private:      // private access specifier
    int y;         // attributes

};

int main() {
  myclass myObj;
  myObj.x = 90;
  //myObj.y = 60;
  return 0;
}


#**18-C++ Encapsulation**

 The meaning of Encapsulation, is to make sure that "sensitive" data is hidden from users. To achieve this, you must declare class variables/attributes as private (cannot be accessed from outside the class). If you want others to read or modify the value of a private member, you can provide public get and set methods.

 **Access Private Members**


To access a private attribute, use public "get" and "set" methods:

**Example explained**
The salary attribute is private, which have restricted access.

The public setSalary() method takes a parameter (s) and assigns it to the salary attribute (salary = s).

The public getSalary() method returns the value of the private salary attribute.

Inside main(), we create an object of the Employee class. Now we can use the setSalary() method to set the value of the private attribute to 50000. Then we call the getSalary() method on the object to return the value.

In [ ]:
%%cu
#include <iostream>
using namespace std;

class Employee {
  private:
    int salary;

  public:
    void setSalary(int s) {
      salary = s;
    }
    int getSalary() {
      return salary;
    }
};

int main() {
  Employee myObj;
  myObj.setSalary(20000);
  cout << myObj.getSalary();
  return 0;
}


#**19-C++ Inheritance**

In C++, it is possible to inherit attributes and methods from one class to another. We group the "inheritance concept" into two categories:

- derived class (child) - the class that inherits from another class
- base class (parent) - the class being inherited from
To inherit from a class, use the : symbol.

In the example below, the Car class (child) inherits the attributes and methods from the Vehicle class (parent):

In [ ]:
%%cu
#include <iostream>
using namespace std;

// Base class
class Vehicle {
  public:
    string brand = "toyota";
    void honk() {
      cout << "Tuut, tuut"<<endl;
    }
};

// Derived class
class Car: public Vehicle {
  public:
    string model = "Mustang";
};

int main() {
  Car myCar;
  myCar.honk();
  cout << myCar.brand + " " + myCar.model;
  return 0;
}



##**19.1-C++ Multilevel Inheritance**

 class can also be derived from one class, which is already derived from another class.

In the following example, MyGrandChild is derived from class MyChild (which is derived from MyClass)

In [ ]:
%%cu
#include <iostream>
using namespace std;

// Parent class
class MyClass {
  public:
    void myFunction() {
      cout << "Some content in parent class." ;
    }
};

// Child class
class MyChild: public MyClass {
};

// Grandchild class
class MyGrandChild: public MyChild {
};

int main() {
  MyGrandChild myObj;
  myObj.myFunction();
  return 0;
}


##**19.2-Multiple Inheritance**

A class can also be derived from more than one base class, using a comma-separated list:

In [ ]:
%%cu
#include <iostream>
using namespace std;

class myclass {
    public:
    void myfunction() {
        cout <<"shahzad hussain"<<endl;
    }
};
class myotherclass {
    public:
    void myotherfunction(){
        cout <<"pakistan"<<endl;
    }
};
class mychildclass:public myclass,public myotherclass{

};
int main() {
    mychildclass myobj;
    myobj.myfunction();
    myobj.myotherfunction();
    return 0;
}

#**19.3-Access Specifiers**

You learned from the Access Specifiers chapter that there are three specifiers available in C++. Until now, we have only used public (members of a class are accessible from outside the class) and private (members can only be accessed within the class). The third specifier, protected, is similar to private, but it can also be accessed in the inherited class:

In [ ]:
%%cu
#include <iostream>
using namespace std;

// Base class
class Employee  {
  protected:  // Protected access specifier
    int salary;
};

// Derived class
class Programmer: public Employee {
  public:
    int bonus;
    void setSalary(int s) {
      salary = s;
    }
    int getSalary() {
      return salary;
    }
};

int main() {
  Programmer myObj;
  myObj.setSalary(70000);
  myObj.bonus = 1500;
  cout << "Salary: " << myObj.getSalary() << "\n";
  cout << "Bonus: " << myObj.bonus << "\n";
  return 0;
}


#**20-Polymorphism**

Polymorphism means "many forms", and it occurs when we have many classes that are related to each other by inheritance.

Like we specified in the previous chapter; Inheritance lets us inherit attributes and methods from another class. Polymorphism uses those methods to perform different tasks. This allows us to perform a single action in different ways.

For example, think of a base class called Animal that has a method called animalSound(). Derived classes of Animals could be Pigs, Cats, Dogs, Birds - And they also have their own implementation of an animal sound (the pig oinks, and the cat meows, etc.):

Remember from the Inheritance chapter that we use the : symbol to inherit from a class.

Now we can create Pig and Dog objects and override the animalSound() method:

In [ ]:
%%cu
#include <iostream>
using namespace std;

class animal{
    public:
    void animalsound(){
        cout <<"animal make a sound"<<endl;
    }
};
class pig:public animal{
    public:
    void animalsound(){
        cout <<"sound of wildforest"<<endl;
    }
};
class dog: public animal{
    public:
    void animalsound(){
        cout <<"sound of gangal"<<endl;
    }
};

int main(){
    animal myanimal;
    pig mypig;
    dog mydog;

    myanimal.animalsound();
    mypig.animalsound();
    mydog.animalsound();
    return 0;

}

#**Reference**
C++ Tutorial
https://www.w3schools.com/cpp/cpp_strings.asp

https://www.learn-cpp.org/en/Variables_and_Types

How to Run CUDA C or C++ on Jupyter (Google Colab)

https://www.wikihow.com/Run-CUDA-C-or-C%2B%2B-on-Jupyter-(Google-Colab)

Code block
http://www.codeblocks.org/downloads/26

Freebootcamp
https://www.freecodecamp.org/

ntroduction to C++ Programming in Urdu
https://www.youtube.com/watch?v=dnCO124Uk1o

6 Sentiment Analysis Projects with Python

https://amankharwal.medium.com/6-sentiment-analysis-projects-with-python-1fdd3d43d90f